<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="robot" >



    <material name="white">
        <color rgba="1 1 1 1" />
    </material>

    <material name="orange">
        <color rgba="1 0.3 0.1 1"/>
    </material>

    <material name="blue">
        <color rgba="0.2 0.2 1 1"/>
    </material>

    <material name="black">
        <color rgba="0 0 0 1"/>
    </material>

    <material name="red">
        <color rgba="1 0 0 1"/>
    </material>
    
    
    
    <!-- Chassis -->


    <xacro:property name="chassis_length" value="0.2"/>
    <xacro:property name="chassis_width" value="0.1"/>
    <xacro:property name="chassis_thickness" value="0.01"/>
    <xacro:property name="front_base_length" value="0.05"/>
    <xacro:property name="back_base_length" value="0.05"/>
    <xacro:property name="a" value="0.02"/>



    <link name="chassis">
		<visual>
            <geometry>
                <box size="${chassis_length} ${chassis_thickness} ${chassis_thickness}"/>
            </geometry>
            <material name="red"/>
            <origin xyz="0 ${-chassis_width/2 +chassis_thickness/2} 0"/>
        </visual>
        <visual>
            <geometry>
                <box size="${chassis_length} ${chassis_thickness} ${chassis_thickness}"/>
            </geometry>
            <material name="red"/>
            <origin xyz="0 ${chassis_width/2 -chassis_thickness/2} 0"/>
        </visual>
        <visual>
            <geometry>
                <box size="${chassis_thickness} ${chassis_width} ${chassis_thickness}"/>
            </geometry>
            <material name="red"/>
            <origin xyz="${-chassis_length/2} 0 0"/>
        </visual>
        <visual>
            <geometry>
                <box size="${chassis_thickness} ${chassis_width} ${chassis_thickness}"/>
            </geometry>
            <material name="red"/>
            <origin xyz="${chassis_length/2} 0 0"/>
        </visual>
        <visual>
            <geometry>
                <box size="${chassis_thickness} ${chassis_width} ${chassis_thickness}"/>
            </geometry>
            <origin xyz="${front_base_length} 0 ${chassis_thickness}"/>
        </visual>
        <visual>
            <geometry>
                <box size="${chassis_thickness} ${chassis_width} ${chassis_thickness}"/>
            </geometry>
            <origin xyz="${-back_base_length} 0 ${chassis_thickness}"/>
        </visual>
    </link>



    




    


    <xacro:property name="l1" value="0.04"/>
    <xacro:property name="l2" value="0.01"/>
    <xacro:property name="l3" value="0.025"/>
    <xacro:property name="thickness" value="0.01"/>
    <xacro:property name="motor_side" value="0.02"/>


    <!-- Left_wheel_base-->

    <link name="left_wheel_base">
        <visual>
            <geometry>
                <box size="${l1} ${thickness} ${thickness}" />
            </geometry>
            <origin xyz="${-l1/2} 0 0"/>
        </visual>
        <visual>
            <geometry>
                <box size="${thickness} ${l2} ${thickness}"/>
            </geometry>
            <origin xyz="${-l1+thickness/2} ${-l2/2-thickness/2} 0.0" rpy="0.0 0.0 0.0"/>
        </visual>
        <visual>
            <geometry>
                <box size="${l3} ${thickness} ${thickness}"/>
            </geometry>
            <origin xyz="${-l1-l3/2} ${-l2} 0.0" rpy="0.0 0.0 0.0"/>
        </visual>
        <visual>
            <geometry>
                <box size="${motor_side} ${motor_side} ${motor_side}"/>
            </geometry>
            <origin xyz="${-l1-l3-motor_side/2} ${-l2} 0.0" rpy="0.0 0.0 0.0"/>
            
        </visual>
    </link>

    <joint name="left_wheel_base" type="continuous">
        <parent link="chassis"/>
        <child link="left_wheel_base"/>
        <origin xyz="${front_base_length} ${chassis_width/2-a} ${chassis_thickness/2}" rpy="0.0 ${-pi/2} 0.0"/>
        
    </joint>

    

    <!-- Left_wheel-->

    <xacro:property name="wheel_radius" value="0.025"/>
    <xacro:property name="wheel_length" value="0.01"/>

    <link name="left_wheel">
        <visual>
            <geometry>
                <box size="${thickness} ${l2-wheel_length/2-motor_side/2} ${thickness}"/>
            </geometry>
            <origin xyz="0 ${-(l2-wheel_length/2-motor_side/2+wheel_length)/2} 0"/>
        </visual>
        <visual>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
            </geometry>
            <origin xyz="0 0 0.0" rpy="${pi/2} 0.0 0.0"/>       
        </visual>

    </link>


    <joint name="left_wheel" type="continuous">
        <parent link="left_wheel_base"/>
        <child link="left_wheel"/>
        <origin xyz="${-l1-l3-motor_side/2} ${0} 0"/>
        <axis xyz="0 1 0"/>
        
    </joint>




    <!-- right_wheel_base-->

    <link name="right_wheel_base">
        <visual>
            <geometry>
                <box size="${l1} ${thickness} ${thickness}" />
            </geometry>
            <origin xyz="${-l1/2} 0 0"/>
        </visual>
        <visual>
            <geometry>
                <box size="${thickness} ${l2} ${thickness}"/>
            </geometry>
            <origin xyz="${-l1+thickness/2} ${-l2/2-thickness/2} 0.0" rpy="0.0 0.0 0.0"/>
        </visual>
        <visual>
            <geometry>
                <box size="${l3} ${thickness} ${thickness}"/>
            </geometry>
            <origin xyz="${-l1-l3/2} ${-l2} 0.0" rpy="0.0 0.0 0.0"/>
        </visual>
        <visual>
            <geometry>
                <box size="${motor_side} ${motor_side} ${motor_side}"/>
            </geometry>
            <origin xyz="${-l1-l3-motor_side/2} ${-l2} 0.0" rpy="0.0 0.0 0.0"/>
            
        </visual>
    </link>

    <joint name="right_wheel_base" type="continuous">
        <parent link="chassis"/>
        <child link="right_wheel_base"/>
        <origin xyz="${front_base_length} ${-chassis_width/2+a} ${chassis_thickness/2}" rpy="0.0 ${-pi/2} ${pi}"/>
        
    </joint>


    <!-- right_wheel-->


    <link name="right_wheel">
        <visual>
            <geometry>
                <box size="${thickness} ${l2-wheel_length/2-motor_side/2} ${thickness}"/>
            </geometry>
            <origin xyz="0 ${-(l2-wheel_length/2-motor_side/2+wheel_length)/2} 0"/>
        </visual>
        <visual>
            <geometry>
                <cylinder radius="${wheel_radius}" length="${wheel_length}"/>
            </geometry>
            <origin xyz="0 0 0.0" rpy="${pi/2} 0.0 0.0"/>       
        </visual>

    </link>


    <joint name="right_wheel" type="continuous">
        <parent link="right_wheel_base"/>
        <child link="right_wheel"/>
        <origin xyz="${-l1-l3-motor_side/2} ${0} 0"/>
        <axis xyz="0 1 0"/>
        
    </joint>



   
	    

    





</robot>
